### PR TITLE
shell.nix: Allow to build on x86 darwin

### PR DESCRIPTION
**Pull request by SevInf**
I've tried to check out qdrant source code and build it on x86 mac laptop, however, i found out a couple of issues with current `shell.nix`.

- Couple of dependencies were missing.
- Default `CXXFLAGS` are not compatible with rocks-db (needs higher deployment target).
- Mold setup did not work with clang out of the box.

This PR fixes all of the issues and allows to build `qdrant` from within nix shell on Intel macOS. I don't have access to ARM Mac at the moment, so I can not say for sure if this configuration will work there.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: shell.nix [hunk 1]
```diff
@@ -60,6 +58,10 @@ mkShellMold {
     pkgs.jq # used in ./tests and ./tools
     pkgs.nixfmt-rfc-style # to format this file
     pkgs.npins # used in tools/nix/update.py
+    pkgs.iconv
+    pkgs.ccache # compile cache for c/cxx code
+    pkgs.sccache # compile cache for rust code
```
Comment (timvisee) [inline shell.nix:63]: I don't think we want this by default. Any specific reason for setting it up here?

Comment (SevInf) [reply shell.nix:63]: It's already used in a shell hook below unconditionally for `RUST_WRAPPER`. 
I can either install nix package on darwin only, remove ccache/sccache usage from shell hook or make it conditional too. What would you prefer?

**Comment by arthurk** [comment]
I've tried a build on this branch using an old intel mac and it worked. Linux works fine without this branch too, so maybe the `iconv` pkg is not necessary there.

File: shell.nix (hunk 1 above)
Comment (xzfc) [reply shell.nix:63]: It was there before, and I prefer not to remove it.

The point of `shell.nix` is not just to have a bare minimum to build Qdrant, but rather to make local development comfortable. So it's a bit opinionated and includes stuff like rust-analyzer, build caching tricks, and stuff to make out git pre-push hooks work.

File: shell.nix [hunk 2]
```diff
@@ -60,6 +58,10 @@ mkShellMold {
     pkgs.jq # used in ./tests and ./tools
     pkgs.nixfmt-rfc-style # to format this file
     pkgs.npins # used in tools/nix/update.py
+    pkgs.iconv
```
Comment (xzfc) [inline shell.nix:61]: Could you please leave a short comment on why `iconv` is needed / which dependency wants it?

Comment (SevInf) [reply shell.nix]: Done

**Review by xzfc — APPROVED** [review]

**Pull request merged**